qemu/tests/migration_with_boot_pattern.py

```diff
@@ -0,0 +1,84 @@
+import aexpect
+import logging
+import time
+
+from autotest.client.shared import utils
+from autotest.client.shared import error
+
+
+@error.context_aware
+def run(test, params, env):
+    """
+    KVM migration test:
+    1) Start VM
+    2) Wait in background for the occurrence of the boot_pattern while
+       migrating the VM in a loop
+    3) Issue one more migration after the pattern is reached
+    4) Optionally send a line to serial port (to proceed booting) followed
+       by yet another migration
+    5) Login to the machine
+
+    :param test: kvm test object.
+    :param params: Dictionary with test parameters.
+    :param env: Dictionary with the test environment.
+    """
+    def wait_for_boot_pattern(boot_pattern, vm):
+        """
+        Waits till last_line_matches and survives vm-migration
+        """
+        patterns = [boot_pattern]
+        while True:
```

Comment:
The thread would not be terminated when `boot_pattern_timeout` expired if really nothing could be got from `vm.serial_console`, then this test would be stuck until `test_timeout` expired.

Reply:
It should terminate after `boot_pattern_timeout` because of the `assert time.time() < end`. The reading (and matching) is done in a different thread so it should not interfere with this loop.